tests/functional/test_auth.py

```diff
@@ -0,0 +1,139 @@
+import unittest
+
+from balena import Balena
+from balena.exceptions import NotLoggedIn, LoginFailed
+from tests.helper import TestHelper
+
+
+class TestAuth(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        cls.balena = Balena()
+
+    def test_01_is_logged_in_false_when_not_logged_in(self):
+        self.balena.auth.logout()
+        self.assertFalse(self.balena.auth.is_logged_in())
+
+    def test_02_logout_does_not_fail_when_not_logged_in(self):
+        self.balena.auth.logout()
+
+    def test_03_not_save_token_with_valid_credentials_on_authenticate(self):
+        TestHelper.load_env()
+        TestAuth.creds = {
+            "username": TestHelper.credentials["user_id"],
+            "password": TestHelper.credentials["password"],
+        }
+
+        token = self.balena.auth.authenticate(**TestAuth.creds)
+        self.assertGreater(len(token), 1)
+
+        # does not save token given valid credentials
+        self.assertFalse(self.balena.auth.is_logged_in())
+
+    def test_04_rejects_with_invalid_credentials(self):
+        with self.assertRaises(LoginFailed):
+            self.balena.auth.authenticate(**{"username": TestAuth.creds["username"], "password": "NOT-CORRECT"})
+
+    def test_05_should_reject_to_get_token_when_not_logged_in(self):
+        self.assertIsNone(self.balena.auth.get_token())
+
+    def test_06_should_be_able_to_login_with_session_token(self):
+        token = self.balena.auth.authenticate(**TestAuth.creds)
+        self.balena.auth.login_with_token(token)
+        self.assertEqual(self.balena.auth.get_token(), token)
+
+    def test_07_should_be_able_to_login_with_api_key(self):
+        key = self.balena.models.api_key.create("apiKey")
```

Comment:
@fisehara I can fix this one but I would like to let you know that these tests are not indepedent. That's why they have the test_{number} prefix. The main reason for that is we could do tests that are fully independent but this slows down our pipeline a lot. Thodoris did the first version where the result of one test can be used on the next one and I agree it is the best we can get here, otherwise deployments get too slow (2+ hours)